autovalet_husky/scripts/find_parking_spot.py

```diff
@@ -0,0 +1,173 @@
+        self.aruco_frame_name = aruco_frame_name
+        self.goal_frame_id    = goal_frame_id
+        self.husky_frame_id   = husky_frame_id
+
+    def collect_tag_poses(self, tag_pose):
+        '''
+        Take in numOfTags tag poses and do RANSAC in the end to avoid outliers
+        '''
+        numOfTags = 20
+        if self.count < numOfTags: # collect tag poses
+            self.aruco_frame_id = tag_pose.header.frame_id
+
+            tf = self.tf_buffer.lookup_transform(self.goal_frame_id, # target_frame_id
+                                    self.aruco_frame_name, # source frame
+                                    rospy.Time(0), # get the tf at first available time
+                                    rospy.Duration(1.0)) # timeout after 1
+            self.tfArray.append(tf)
+            self.count += 1
+        elif self.count == numOfTags: # perform RANSAC on transformArray
+            self.tag_tf = self.tag_pose_RANSAC()
+            self.pub_two_goals()
+            self.count += 1
+
+    def pub_two_goals(self):
+        parking = False
+        pos1 = [0, 2, 1]
+        rot1 = [-np.pi/2, 0, -np.pi/2]
+        goal1 = self.generate_parking_goal(self.tag_tf, pos1, rot1)
+        while not parking:
+            rospy.sleep(0.5)
+            # keep publishing the first goal until the robot is close enough
+            if self.dist_to_first_goal(goal1) > 0.5:
+                self.pub.publish(goal1)
+                print("first goal published")
+            else:
+                pos2 = [0, 0, 4]
+                rot2 = [0, -np.pi/2, -np.pi/2]
+                goal2 = self.generate_parking_goal(self.tag_tf, pos2, rot2)
+                self.pub.publish(goal2)
+                print("second goal published")
+                parking = True
+        return
+
+    def dist_to_first_goal(self, goal1):
+        base_link_tf = self.tf_buffer.lookup_transform(self.goal_frame_id, # target_frame_id
+                                self.husky_frame_id, #source frame
+                                rospy.Time(0), #get the tf at first available time
+                                rospy.Duration(1.0)) #timeout after 1
+        base_link_x = base_link_tf.transform.translation.x
+        base_link_y = base_link_tf.transform.translation.y
+        first_goal_x = goal1.pose.position.x
+        first_goal_y = goal1.pose.position.y
+        dist = np.linalg.norm(np.array([base_link_x - first_goal_x, base_link_y - first_goal_y]))
+
+        return dist
+
+    def tag_pose_RANSAC(self):
+        quatArray = []
+        for i in range(len(self.tfArray)):
+            quat = self.tfArray[i].transform.rotation
+            quatArray.append(np.array([quat.x, quat.y, quat.z, quat.w]))
+        votes = []
+        for i in range(len(quatArray)):
+            vote = 0
+            for j in range(len(quatArray)):
+                diff = np.abs(1 - np.dot(quatArray[i], quatArray[j]) ** 2)
```

Comment:
had no idea you could take differences of quaternions. cool stuff

Reply:
can u link where u found this formula

Reply:
done